src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleLogFormatter.cs

```diff
@@ -0,0 +1,168 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Extensions.Logging.Console
+{
+    internal class JsonConsoleLogFormatter : ConsoleFormatter, IDisposable
+    {
+        private IDisposable _optionsReloadToken;
+
+        public JsonConsoleLogFormatter(IOptionsMonitor<JsonConsoleFormatterOptions> options)
+            : base (ConsoleFormatterNames.Json)
+        {
+            FormatterOptions = options.CurrentValue;
+            ReloadLoggerOptions(options.CurrentValue);
+            _optionsReloadToken = options.OnChange(ReloadLoggerOptions);
+        }
+
+        public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeProvider scopeProvider, TextWriter textWriter)
+        {
+            LogLevel logLevel = logEntry.LogLevel;
+            string category = logEntry.Category;
+            int eventId = logEntry.EventId.Id;
+            string message = logEntry.Formatter(logEntry.State, logEntry.Exception);
+            Exception exception = logEntry.Exception;
+            const int DefaultBufferSize = 1024;
+            var output = new ArrayBufferWriter<byte>(DefaultBufferSize);
+            using (var writer = new Utf8JsonWriter(output, FormatterOptions.JsonWriterOptions))
+            {
+                writer.WriteStartObject();
+
+
+                string timestamp = null;
+                var timestampFormat = FormatterOptions.TimestampFormat;
+                if (timestampFormat != null)
+                {
+                    var dateTime = FormatterOptions.UseUtcTimestamp ? DateTime.UtcNow : DateTime.Now;
+                    timestamp = dateTime.ToString(timestampFormat);
+                }
+                writer.WriteString("timestamp", timestamp);
+                writer.WriteNumber("eventId", eventId);
+                writer.WriteString("logLevel", GetLogLevelString(logLevel));
+                writer.WriteString("category", category);
+                writer.WriteString("message", message);
+
+                if (exception != null)
+                {
+                    writer.WriteStartObject("exception");
+                    writer.WriteString("message", exception.Message.ToString());
+                    writer.WriteString("type", exception.GetType().ToString());
+                    writer.WriteStartArray("stackTrace");
+                    if (exception?.StackTrace != null)
```

Comment:
nit: assign `exception?.StackTrace` to a local so that we can lose the nullability checks below.

Reply:
exception is already checked for null before so we can just remove '?'

---
In reply to: [451146045](https://github.com/dotnet/runtime/pull/38616#discussion_r451146045) [](ancestors = 451146045)